### ISSUE TITLE
gradient of normal distribution negative log-likelihood

### ISSUE THREAD

**Issue by guyko81** [stanfordmlgroup/ngboost#6]
I might be wrong but the gradient seems incorrect wrt sigma

I think this should be used, my tests give bad results on boston (nll) with the current implementation, with the following change it's good again

```
    def D_nll(self, Y_):
        Y = Y_.squeeze()
        D = np.zeros((self.var.shape[0], 2))
        D[:, 0] = (Y - self.loc) / self.var
        D[:, 1] = (Y - self.loc)**2 / self.scale**3 - 1/self.scale
        return -D
```

**Comment by avati** [comment]
Hi @guyko81,

Thank you for taking a close look. I checked the math once more, and I think the code looks correct for me. Note that our parametrization is not (mu, sigma), but rather (mu, log-sigma). Your answer looks correct for (mu, sigma) parametrization.

In order to explain the NLL performance on Boston housing dataset - have you set a fixed random seed so that multiple runs are comparable? Otherwise the comparison between the two gradients could be confounded with random noise.

**Comment by guyko81** [comment]
Hi @avati,

I see your point, didn't realise this. However the tests show that for some reason the scale parameter seems to be working better (I copy my code maybe you spot some mistake).

I have 2 runs with 20 iterations with each settings. Validation set is 20% (I used the example code from the readme.md). DecisionTreeRegressor(max_depth=3), n_estimators=100, learning_rate=0.1

```
D[:, 0] = (self.loc - Y) / self.var
D[:, 1] = 1 - ((self.loc - Y) ** 2) / self.var
return D
```
RMSE:  mean: 3.07741915385 std: 0.480625201044
NLL:  mean: 4.21798940883 std: 1.08163088748

and 

RMSE:  mean: 3.00950118802 std: 0.371132936142
NLL:  mean: 4.48953001556 std: 1.19808307636

and with 

```
D[:, 0] = (Y - self.loc) / self.var
D[:, 1] = (Y - self.loc)**2 / self.scale**3 - 1/self.scale
return -D
```

RMSE:  mean: 3.02053535936 std: 0.490099157414
NLL:  mean 2.68636672134 std: 0.0453827523916

and

RMSE:  mean: 3.11187436877 std: 0.555482856713
NLL:  mean: 2.68262090161 std: 0.0661654897441

I was testing with random forest too, the result is more smooth (an idea: you could show that this approach with random forest can produce even better results compared to normal trees and the neural network approaches).

```
X, Y = load_boston(True)

RMSE = []
NLL = []
for k in range(20):
    print(k)
    X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2)
    
    tree_learner_with_depth = lambda d: DecisionTreeRegressor(
    criterion='friedman_mse',
    min_samples_split=2,
    min_samples_leaf=1,
    min_weight_fraction_leaf=0.0,
    max_depth=d,
    splitter='best')
    base_tree_learner = lambda: tree_learner_with_depth(3)
    
    #rf = lambda d: RandomForestRegressor(n_estimators=20, max_depth=d, verbose=0, criterion='friedman_mse')
    #rf_base_model = lambda: rf(5)
    ngb = NGBoost(n_estimators=100, learning_rate=0.1, Base=base_tree_learner, Dist=Normal, Score=MLE(), natural_gradient=True,
                  verbose=False)
    ngb.fit(X_train, Y_train)
    Y_preds = ngb.predict(X_test)
    Y_dists = ngb.pred_dist(X_test)
    # test Mean Squared Error
    RMSE.append(np.sqrt(mean_squared_error(Y_preds, Y_test)))
    #test Negative Log Likelihood
    NLL.append(-Y_dists.logpdf(Y_test.flatten()).mean())
```

**Comment by avati** [comment]
@guyko81 

I see that with the current code, there is quite a lot of variance in the NLL estimates (std of ~ 1.08 and 1.19) whereas the estimates are lot sharper for the second case. Perhaps it just means that with this parametrization the model needs a different set of hyperparameters? (learning rate and/or n_estimators)

**Comment by guyko81** [comment]
Hi @avati 
Actually I see the mistake! It's so annoying - the negative sign: 
instead of `(self.loc - Y)` we need `(Y - self.loc)`
```
        D[:, 0] = (self.loc - Y) / self.var
        D[:, 1] = 1 - ((Y - self.loc) ** 2) / self.var
        return D
```

**Comment by avati** [comment]
I checked again. I believe the current code is proper. I also tried switching the sign and it resulted in a LOT worse performance on the benchmark tasks.

**Comment by avati** [comment]
Can you come up with a standalone and reproducible script that shows your version of the log-sigma gradient gives better NLL on validation set?

**Comment by guyko81** [comment]
Yes, I create one. I have tested on another domain and just feels better. I see the math that it shouldn't be better but still..
Should I email you or copy here is enough?

**Comment by avati** [comment]
You can email me at avati at cs dot stanford dot edu

**Comment by avati** [comment]
Based on the discussion on email (summary): it appears that @guyko81  version of the gradient results in smaller gradient values for the log-sigma parameter, but it seems like the real benefit of it was an effectively smaller learning rate due to the downscaling. Just using a smaller overall learning rate gave even better performance since the mean and log-sigma updates are better balanced.

Closing this issue for now. If any new observations are found, please file a new issue.
